### PR TITLE
accounts/abi/bind: link dependent libs in deploy

### PR DESCRIPTION
**Pull request by gballet**
This PR addresses https://github.com/ethereum/go-ethereum/issues/17010

Upon loading the `solc` output, `abigen` will make a list of potential "linker" references to pass the template. If a linker reference is found, a call to the library deployment code will be added to the contact deployment code, and the bytecode will be updated with the deployment address.

I also had to change the way bind was being tested: now the test has to specify all dependent ABIs and bytecodes, as well as the list of contract dependencies. Basically, this assumes an input a bit closer to what `abigen` provides to the `Bind` function.

### CODE REVIEW

**Comment by gballet** [comment]
@ligi could you please have a look at the java code being generated?

**Comment by gballet** [comment]
Didn't mean to close this :man_facepalming:

**Comment by rjl493456442** [comment]
A few changes I made 

```go
diff --git a/cmd/abigen/main.go b/cmd/abigen/main.go
index 16cb0df08..ab70b6f46 100644
--- a/cmd/abigen/main.go
+++ b/cmd/abigen/main.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"regexp"
 	"strings"
 
 	"github.com/ethereum/go-ethereum/accounts/abi/bind"
@@ -82,7 +83,7 @@ func main() {
 		bins  []string
 		types []string
 		sigs  []map[string]string
-		libs  map[string]string
+		libs  = make(map[string]string)
 	)
 	if *solFlag != "" || *vyFlag != "" || *abiFlag == "-" {
 		// Generate the list of types to exclude from binding
@@ -151,7 +152,19 @@ func main() {
 				os.Exit(-1)
 			}
 		}
-		bins = append(bins, string(bin))
+		// If there is any library references, catch them.
+		segments := strings.Split(string(bin), "//")
+		for i := 1; i < len(segments); i++ {
+			segment := strings.TrimSpace(segments[i])
+			parts := regexp.MustCompile(`\$([0-9a-f]*)\$ -> (.*)`).FindStringSubmatch(segment)
+			if len(parts) != 3 {
+				fmt.Println("Failed to parse library reference")
+				os.Exit(-1)
+			}
+			nameParts := strings.Split(parts[2], ":")
+			libs[parts[1]] = nameParts[len(nameParts)-1]
+		}
+		bins = append(bins, strings.TrimSpace(segments[0]))
 
 		kind := *typFlag
 		if kind == "" {

```

But we can't get the library bin from the input, how can we deploy the library first?

File: cmd/abigen/main.go [hunk 1]
```diff
@@ -128,6 +130,9 @@ func main() {
```
Comment (rjl493456442) [inline cmd/abigen/main.go]: Regarding the `bin` here. It should have some additional information like `//$812de29bd0a57d8fddaebd9e2f77569c6c$ -> /solidity/lib.sol:Set`. Let's filter them out first.

Comment (rjl493456442) [reply cmd/abigen/main.go]: It only exists in bin file. Never mind.

**Comment by rjl493456442** [comment]
The golang part is good to me. But I try the java binding, get these errors

```
bytecode.replace("__$812de29bd0a57d8fddaebd9e2f77569c6c$__", setInst.Address);
```

Error: can't resolve method `replace(java.lang.String, org.etherem.geth.Address)`

And also if we have multi contracts(e.g. contract + library), we will have 2 java public classes which can be defined in single java class file.

**Pull request merged**